metricbeat/module/elasticsearch/node_stats/data_xpack.go

```diff
@@ -0,0 +1,225 @@
+			"max_file_descriptors":  c.Int("max_file_descriptors"),
+			"cpu": c.Dict("cpu", s.Schema{
+				"percent": c.Int("percent"),
+			}),
+		}),
+		"jvm": c.Dict("jvm", s.Schema{
+			"mem": c.Dict("mem", s.Schema{
+				"heap_used_in_bytes": c.Int("heap_used_in_bytes"),
+				"heap_used_percent":  c.Int("heap_used_percent"),
+				"heap_max_in_bytes":  c.Int("heap_max_in_bytes"),
+			}),
+			"gc": c.Dict("gc", s.Schema{
+				"collectors": c.Dict("collectors", s.Schema{
+					"young": c.Dict("young", s.Schema{
+						"collection_count":          c.Int("collection_count"),
+						"collection_time_in_millis": c.Int("collection_time_in_millis"),
+					}),
+					"old": c.Dict("young", s.Schema{
+						"collection_count":          c.Int("collection_count"),
+						"collection_time_in_millis": c.Int("collection_time_in_millis"),
+					}),
+				}),
+			}),
+		}),
+		"thread_pool": c.Dict("thread_pool", s.Schema{
+			"bulk": c.Dict("bulk", s.Schema{
+				"threads":  c.Int("threads"),
+				"queue":    c.Int("queue"),
+				"rejected": c.Int("rejected"),
+			}),
+			"generic": c.Dict("generic", s.Schema{
+				"threads":  c.Int("threads"),
+				"queue":    c.Int("queue"),
+				"rejected": c.Int("rejected"),
+			}),
+			"get": c.Dict("get", s.Schema{
+				"threads":  c.Int("threads"),
+				"queue":    c.Int("queue"),
+				"rejected": c.Int("rejected"),
+			}),
+			"index": c.Dict("index", s.Schema{
+				"threads":  c.Int("threads"),
+				"queue":    c.Int("queue"),
+				"rejected": c.Int("rejected"),
+			}),
+			"management": c.Dict("management", s.Schema{
+				"threads":  c.Int("threads"),
+				"queue":    c.Int("queue"),
+				"rejected": c.Int("rejected"),
+			}),
+			"search": c.Dict("search", s.Schema{
+				"threads":  c.Int("threads"),
+				"queue":    c.Int("queue"),
+				"rejected": c.Int("rejected"),
+			}),
+			"watcher": c.Dict("watcher", s.Schema{
+				"threads":  c.Int("threads"),
+				"queue":    c.Int("queue"),
+				"rejected": c.Int("rejected"),
+			}),
+		}),
+		"fs": c.Dict("fs", s.Schema{
+			"summary": c.Dict("total", s.Schema{
+				"total_in_bytes":     c.Int("total_in_bytes"),
+				"free_in_bytes":      c.Int("free_in_bytes"),
+				"available_in_bytes": c.Int("available_in_bytes"),
+			}),
+		}),
+	}
+)
+
+func eventsMappingXPack(r mb.ReporterV2, m *MetricSet, content []byte) {
+	nodesStruct := struct {
+		ClusterName string                            `json:"cluster_name"`
+		Nodes       map[string]map[string]interface{} `json:"nodes"`
+	}{}
+
+	json.Unmarshal(content, &nodesStruct)
+
+	// Normally the nodeStruct should only contain one node. But if _local is removed
+	// from the path and Metricbeat is not installed on the same machine as the node
+	// it will provid the data for multiple nodes. This will mean the detection of the
+	// master node will not be accurate anymore as often in these cases a proxy is in front
+	// of ES and it's not know if the request will be routed to the same node as before.
```

Comment:
In order to support proxies, we _could_ [allow users to specify the node name](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster.html#cluster-nodes) that they want Beats to monitor and, if unspecified, use `_local` as the name.

Pitfalls:

- Misconfiguration means it's broken.
- Duplicate node names can cause confusion.
- Requires the module's configuration to be reloaded if the node is renamed.

Advantages:

- We'll work, by default, for non-proxies.
- We can work through proxies.

Reply:
That is pretty interesting. I would skip it for the first iteration but I'm pretty sure that is a request that will pop up.

It will probably be a bit tricker for dynamic environments where the user does not have much control around the naming of nodes.

Reply:
In that case they can set its address, which hopefully they know.